src/Http/Cookie/Cookie.php

```diff
@@ -0,0 +1,468 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.5.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http\Cookie;
+
+use Cake\Chronos\Chronos;
+use Cake\Utility\Hash;
+use DateTimeInterface;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Cookie object to build a cookie and turn it into a header value
+ *
+ * An HTTP cookie (also called web cookie, Internet cookie, browser cookie or
+ * simply cookie) is a small piece of data sent from a website and stored on
+ * the user's computer by the user's web browser while the user is browsing.
+ *
+ * Cookies were designed to be a reliable mechanism for websites to remember
+ * stateful information (such as items added in the shopping cart in an online
+ * store) or to record the user's browsing activity (including clicking
+ * particular buttons, logging in, or recording which pages were visited in
+ * the past). They can also be used to remember arbitrary pieces of information
+ * that the user previously entered into form fields such as names, addresses,
+ * passwords, and credit card numbers.
+ *
+ * @link https://tools.ietf.org/html/rfc6265
+ * @link https://en.wikipedia.org/wiki/HTTP_cookie
+ */
+class Cookie implements CookieInterface
+{
+
+    use CookieCryptTrait;
+
+    /**
+     * Cookie name
+     *
+     * @var string
+     */
+    protected $name = '';
+
+    /**
+     * Raw Cookie value
+     */
+    protected $value = '';
+
+    /**
+     * @var bool
+     */
+    protected $isExpanded = false;
+
+    /**
+     * Expiration time
+     *
+     * @var int
+     */
+    protected $expiresAt = 0;
+
+    /**
+     * Path
+     *
+     * @var string|null
+     */
+    protected $path = null;
+
+    /**
+     * Domain
+     *
+     * @var string|null
+     */
+    protected $domain = null;
+
+    /**
+     * Secure
+     *
+     * @var bool
+     */
+    protected $secure = false;
+
+    /**
+     * HTTP only
+     *
+     * @var bool
+     */
+    protected $httpOnly = false;
+
+    /**
+     * Constructor
+     *
+     * @param string $name Cookie name
+     * @param string|array $value Value of the cookie
+     */
+    public function __construct($name, $value = '')
```

Comment:
Should the constructor support all the properties that could be set? It might make the constructor long, but would save you some method calls when the entire cookie state is known at construction time.